### ISSUE TITLE
Enhancement: [naming-convention] don't error for object keys in typed objects

### ISSUE THREAD

**Issue by simon-abbott** [typescript-eslint/typescript-eslint#6120]
### Before You File a Proposal Please Confirm You Have Done The Following...

- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22enhancement%3A+plugin+rule+option%22) and found none that match my proposal.
- [X] I have searched the [current rule list](https://typescript-eslint.io/rules/#supported-rules) and found no rules that match my proposal.
- [X] I have [read the FAQ](https://typescript-eslint.io/docs/linting/troubleshooting) and my problem is not listed.

### My proposal is suitable for this project

- [X] I believe my proposal would be useful to the broader TypeScript community (meaning it is not a niche proposal).

### Link to the rule's documentation

https://typescript-eslint.io/rules/naming-convention

### Description

I have several projects that are using `@typescript-eslint/naming-convention` to enforce camelCase properties in object literals (among other places). However, there are quite a few libraries out there that take options objects with snake_case keys as part of their API. This means that any time I want to specify one of those options I need to `eslint-disable` the `naming-convention` rule, otherwise it errors.

My proposal is to either add an extra flag (or a separate rule, but it's really an extension of `naming-convention`) that looks at the type information of the object literal being defined, and skips over any keys that are explicitly defined in that type. Type definitions, generic/typeless object literals, and unknown keys should still be subject to validation, just not known keys in explicitly typed object literals.

### Fail

```typescript
// With format set to not allow 'snake_case'

const someObj = {
  bad_key: 'hello',
};

interface SomeObjectType {
  something_here: string;
}

type MapType: Record<string, unknown>;
const someMap: MapType = {
  // The type does not dictate that this key must be snake_case,
  // so it should error.
  bad_key: true;
}
```

### Pass

```typescript
// With format set to not allow 'snake_case'

interface SomeObjectType {
  // eslint-disable-next-line @typescript-eslint/naming-convention
  something_here: string;
}

const anotherObject: SomeObjectType = {
  // This one should not error, since this format is _required_ for the object
  // to match the type, and so the programmer here has no choice but to use the
  // bad syntax.
  something_here: 'no error!',
};

function doSomething(obj: SomeObjectType) {}

// This should also pass since the key format is outside the caller's control.
doSomething({ something_here: 'just works!' });

type CombinedType = SomeObjectType & Record<string, unknown>;
const complex: CombinedType = {
  // Even though CombinedType has a generic `[key: string]` typing, it also
  // explicitly has the key `something_here`, which should be allowed.
  something_here: 'still works',
}
```

**Comment by bradzacher** [comment]
But what is a "typed object"?
How can you tell that an object is "typed"?

---

Have you tried ignoring the property names instead of disabling them ad-hoc? This example has config to ignore specific names:
https://typescript-eslint.io/rules/naming-convention#ignore-properties-that-require-quotes

**Comment by simon-abbott** [comment]
> But what is a "typed object"? How can you tell that an object is "typed"?

A "typed object" is one where the shape of the object is defined outside of the object literal. Rather than the type being inferred / generated _from_ the object literal, the object literal must conform to an existing type. It's a clear distinction, but I have no idea how easy it is to get that information out of whatever type API you are using, or if it's even possible.

> Have you tried ignoring the property names instead of disabling them ad-hoc? This example has config to ignore specific names: https://typescript-eslint.io/rules/naming-convention#ignore-properties-that-require-quotes

That is technically possible, yes, but we have a large codebase that interfaces with many 3rd party libraries, and I don't want to issue a blanket "this name is ok" since we might use that same name in our own internal code somewhere else, and we'd want it to be camelCase there. The only other option then would be to define those exceptions on a per-directory or per-file basis, but that's even _more_ verbose than just using `eslint-ignore`.

**Comment by JoshuaKGoldberg** [comment]
👍 yeah I've been hit by this before. It's really unfortunate when external libraries don't adhere to common JS naming conventions, like `docker.createContainer` in that example. But even if they do, they might not be the same idioms as what you use.

**Comment by bradzacher** [comment]
Yeah from my experience most codebases use camel case as do most libraries.

So the mismatches you have are:
- codebases that use camel case (common) interfacing with libraries that don't accept camel case (rare - aws and docker for example)
- codebases that don't use camel case (rare) interfacing with libraries that do (common)

For the second case - that's where you'd really want this sort of feature. But it's a small subset of our user base that uses that style - so there's very little motivation there.

For the first (at least with docker and aws) the interaction surface area is often pretty small relative to the entire codebase size - meaning its either feasible to allowlist the few names needed, or its palatable enough to use disable comments on the dozen or so lines. So there's not a huge amount of motivation there either. People kind of "make do" and move on.

If we could cheaply and quickly calculate types I would do this in a heartbeat. But as it stands naming-convention is already one of the slowest rules on large codebases due to the sheer volume of things it checks the AST for - so adding types is going to exacerbate that a lot.
Anecdotal perf side note - it's the slowest rule but 5x (about 30s - 15% of the lint time) on the internal Canva codebase which is only configured with the `typeLike` selector! Perf is an open todo for me to look into.

**Comment by simon-abbott** [comment]
> [T]he only solution is to pull in type info for the rule so that you can correctly catch the typed vs untyped difference. _**Adding type info here would be a breaking change.**_

I am OK with this feature being a breaking change, though I understand why that would be considered a negative. I would also be fine with adding a new rule/option that enables type checking, but that could easily cause problems since having an option enable type checking probably isn't easy to implement (or understand), and having a separate rule could easily lead to linting every token twice, if someone enabled both.

> Considering this rule checks every single property in the codebase - that's going to be a pretty expensive type calculation!

Yeah, with how slow the TS engine is, that's a pretty compelling argument against this proposal.

> Also worth noting that even with types we can't catch cases like this:
> 
> ```ts
> const config = {
>   Image: image,
>   Cmd: command,
> };
> const dockerContainer = docker.createContainer(config); 
> ```
> 
> In order for the new checks to catch this case you'd need to explicitly annotate the variable (note that an `as const` wouldn't work either!).

That's exactly how I would expect that situation to work. That's a standalone ("untyped") object that just happens to be passed to `createContainer`.


> Its a really difficult problem to solve well and there aren't a huge amount of bad APIs out there (aws and docker are the two I know of) - meaning it is the exception to the rule, not a super common case across most of our user base.

> Yeah from my experience most codebases use camel case as do most libraries.

> For the first [...] the interaction surface area is often pretty small relative to the entire codebase size [...] [so] its palatable enough to use disable comments on the dozen or so lines.

Our codebase enforces camelCase, but we use several libraries that require snake_case keys. The number is small enough that we definitely can get by using disable comments, but it's still just common enough that it's annoying.

> If we could cheaply and quickly calculate types I would do this in a heartbeat. But as it stands naming-convention is already one of the slowest rules on large codebases due to the sheer volume of things it checks the AST for - so adding types is going to exacerbate that a lot.

Maybe ~~when~~ if [stc](https://github.com/dudykr/stc) ever becomes stable it will be fast enough to offer this feature, but I won't hold my breath. 😞

**Comment by simon-abbott** [comment]
It's also worth noting that while doing the actual type checking would indeed be expensive, we _could_ reduce the number of places that would be necessary. Variables, functions, and parameters (basically anything outside of an object literal) will _never_ be typed, which can be determined based only on the untyped AST, types only need to even be considered for freestanding objects that have an explicit type (either `const foo: SomeType = {}`, or `const foo = {} satisfies SomeType`) and object literals that are created in a potentially typed context.

For example:
```ts
// These can be analyzed correctly without types
const someVar = 123;
function foo(arg1: number): string {}
const someObj = { foo: 'bar' };
someFunc(someObj);

// These need type information to make sure
const typedObj: SomeType = { foo: 'bar' };
const satisfiedObj = { foo: 'bar' } satisfies SomeType;
someFunc({ foo: 'bar' });
```

The combination of the breaking change + the performance hit is still quite a compelling argument against this change at the moment, but it's not as universally bad as it might appear. I don't have metrics on how often things fall into those two categories though, so I can't guess what the actual performance impact would be.

**Comment by sharky98** [comment]
I was redirected here by Josh Cena on my bug issue I opened earlier #7109. I closed it as duplicate, since the root cause is the same (object that are typed).

While the discussion above mostly revolve around external library using camelCase or not, my use case I wanted to use a naming convention regarding boolean having a verb before (like the one shown in the examples).

As for a solution, it seems prior discussion tried to go with using typing and AST to figure out when to ignore or not the rule. I have next to zero knowledge in the inner of TS, ESLint and such, but when looking at the source code of `naming-convention`, it seems to be using `ESTree`. Looking at my [updated playground] for my case, I was wondering if some basic condition could be used?

* Is the `ObjectExpression` part of a `VariableDeclarator` or `CallExpression`?
   * NO => Run the `naming-convention` rule.
   * YES `VariableDeclarator` => Does the `id: Identifier {}` contain a `typeAnnotation` field?
      * NO => Run the `naming-convention` rule.
         * Logic: we are creating this object without restriction, so we apply our rules.
      * YES => Do **NOT** run the `naming-convention` rule.
         * Logic: we are creating this object with type restriction, so we do not apply our rules.
   * YES `CallExpression` => Do **NOT** run the `naming-convention` rule.
      * Logic: inside a call expression (typed or not), the called is expecting specific keys.

Like I said, I have next to zero knowledge in the inside of TS and ESLint. So maybe your discussion about AST and added expensive type calculation was already covering my vague idea above.

[playground]:https://typescript-eslint.io/play/#ts=5.0.4&sourceType=module&showAST=es&code=MYewdgzgLgBAtgTwLIFMoAsQBMYF4YAUADgE4hEQBcMA3jEggEIggA2KAhmNQEYvtcA3DACWEBszaduMPlKEwAvgEo8APloBYAFAwYAen17jJ02ZMA9GAAEoCIigjASIolAC0j1iLBR9YDjgfAHN3UDAANxRfEXAYKBdg4JQSCBgONJQADwdgKBQsADodPXCIKULWEGDiMgplQR1FRu0dOwd6BAAVexQ8LV1OyQEZORGWvUMYK1tepxc3TwhvX39AkLDwKJi4hJEklLSMmGzc-KKS0XEmfmleW6Emlp0y2EQAeQB3MHeeACsUHlqAweh18DRLhIHjIEgBXFAAGkuUxm7UczlcHi8Pj8ASCYFC4W2UFiYHiiWSqQRslhsAAkuk4DAAGZsKqfELxdB9ABErJIwAKPPi9hCxUGYih8hkzI4rAgKCeOhe4Gg8AQXx+-0BUAAciAoKCCv0IYMpSNqHDEcijKi5hjFtjVniNkToiTdhTDulMjkdQVxXpJTdpdRZfLFdpmsrtIhUBhsAQ6Oa7iy5QrqcHhqmrUoGjopuZpjY0fNMUsVrj1gTNpF3aTyftKRBqTxaTAGYEWWyQByCVzefzBVhhe0xUA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6JgQwFtmBzWsgHsmAN0QtKw9GADa4bDgDuHaE0gAaeQqwLskZIiRl8g6NMjFog0tCIatuyEVKoMc3Y4BGgwUg5qHBQBdTQ8cADNTLg58aXcwnAAFDmQyDngAYRTESEDsELycS0RwygAPOMK9SlRQhKhkAAtBBAATe3qoRpSO+sg0tTqEyFbKdqGwyEVKeHhchKC8gF8HRYgVpaA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA

**Comment by simon-abbott** [comment]
@sharky98 That would technically work, but only if every annotated variable and called function specifies specific types.

For example, what should happen with the following code?

```ts
function doSomething(obj: Record<string, any>): void { /* snip */ }
doSomething({ bad_casing: 'oh no!' });
```

Or how about my example from the original post?

```ts
interface SomeObjectType {
  // eslint-disable-next-line @typescript-eslint/naming-convention
  something_here: string;
}

type CombinedType = SomeObjectType & Record<string, unknown>;

const complex: CombinedType = {
  // This should be allowed
  something_here: 'still works',
  // But this should not
  bad_casing: 'is wrongly allowed',
}
```

Using the logic you described, both of these examples would pass linting, which is not what I would expect.

However, if we tweak your logic from "don't run `naming-convention`" to "run `naming-convention` with types enabled", then your logic is basically exactly what I was trying to describe [here](https://github.com/typescript-eslint/typescript-eslint/issues/6120#issuecomment-1397632112). It's still a performance hit, but at least it's not as bad as it would be if types were enabled for _every_ object.

Another potential optimization would be to only check the types if it matches the above logic _and_ fails the initial check. So for example:

```ts
interface TypeA {
  // eslint-disable-next-line @typescript-eslint/naming-convention
  something_here: string;
}

interface TypeB {
  somethingHere: string;
}

// This requires a type check, because "something_here" is not camelCase,
// so we need to check if there is an external constraint.
const thingA: TypeA = { something_here: 'this is fine' };

// This requires no type check, because "somethingHere" is already camelCase, we don't care why.
const thingB: TypeB = { somethingHere: 'also fine!' };
```

**Comment by sharky98** [comment]
Thanks for the feedback! I clearly didn't think about the usage of `Record<string,any>`. However, I would tend to disagree with the proposal you made on how the rule should deal with this case.

As you put it `The type does not dictate that this key must be snake_case, so it should error.`. Yet, if the callee typed an argument as a `Record<string,any>`, my understanding is that the callee intended the keys to be anything; otherwise it should have been typed with an object literal.

For sure, there would be rules in places on how the called would parse the keys, noted in the documentation somewhere. However, I am not sure how a linter would ever know what case the callee is expecting in his `Record<string,any>`. The keys could be in any casing `snake_case`, `PascalCase`, `kebab-case` or even includes non-alpha characters!

I think a good example of this situation is `typescript-eslint` itself! The type for the configuration of the rules is `export type RulesRecord = Partial<Record<string, RuleEntry>>;`. Yet, the key expected for the `Record<>` are rule names, which are "pluggable" and written in `kebab-case` with additionnal non-alpha characters (such as "@typescript-eslint/naming-convention")[^1].

https://github.com/typescript-eslint/typescript-eslint/blob/082c948cf68b2b17212144b62608962f0fcd0069/packages/utils/src/ts-eslint/Linter.ts#L121

With that said, I am looking at the ESTree shown in the playground linked below and I think (again, *think*😅) that my logic still stand.

[^1]:Key like those would be in quotes. However, the rule still check in quotes, unless otherwise specified with a rule like [this one](https://typescript-eslint.io/rules/naming-convention/#ignore-properties-that-require-quotes).

## TL;DR

One cannot assume `Record<string,any>` has to be `camelCase` (or whatever the configuration says).

## [Playground](https://typescript-eslint.io/play/#ts=5.0.4&sourceType=module&code=PTAEAUBsEME8HMBOB7ArgOwCagGbMaAMQBsAjAEwAMAsAFAigAWALswA4DOAXCPAJbNGqAEYA6AMbIAtsGaw2AUw7jEfNswC0SyH3TNZ8pSrWbtu-Xw4dUS4GSp06IQoVAAxAIIBJADJPggKDkoAC8AHygAKKIKIgK2NAcoAoAHorizHGO9GBBACIKOLpxTAqxdJLoHMygUrAAsgqCyNjBoAAUbCicXKAA3qB1sABCyMiQCtDoPcKj45MA3KCWgyNjE1OgM2sLoAC+AJQh4b10oKAVHGuikMjwHV0c+-N0u8+0dHKKA7AAKoYhfVOoAYeQKRWwjFKCiBK1m62mcIWQOWw0RGy2c3Qb1eWXMpRw0HEClAAGVpAoAPLCABWCnSfy+J1oZxBoHyhXQxUhZWZoEuUkajF08AA+tyFD0qqp0PBsVkLtVahSAO7oKm09IAOWQzAZxVaTJZOTZYM5EKhoGVfBYaGqnyUMNR2w2zEQNgANMiOLDnT0CZAONDaDj3rQFXzyeqAYbgcb2eCSrFLdbmLbQPaOEDhNBMCKANYKWA9ADkkMgN2LnuDb38LlJAAkKQBVHy5CAeEkk-ygQC8G4BSnaOpMYaEg2HQOs2xKiMUyoYY-dATY4wtA0FAmFNxTaKQyiHQ0Egh3t7tAtuXG9AnVuiGgUgFiGLiQqrrGonKyEqitgKrVNLpzB6QY9WjR1VkxHpXQ9L0fXA3AD0DF4a2yXsByXFc1w3Dktx3Up90PdNDBPM8+AvK8kFve9H3OD8X0gN8ww-KpV3HQRSnVf8ejJAV2Ppf4DSBBgfiFRIP2JDhh1QUdQBYpJonwE9l3QIl02E3B8CkaBqksUARViABHVA+FiXM1IIVjTz-dIBLAVMak08RGBU4lj2Y7BLicy8UHIu9SkTYlGASaTkHOYc+GU4RUDtYLUEDJzrM2HM+VgPRoGSeizn5QVhTFKES3HWSYgAQkrRCsnnVDFPgVd103bBt2SXc8KPQjT0i89iXEA9xgfJ8aJQOi6FqBomkwNp+hg+E4IDBQTxRMDJsg4kDiQ8rF0q6rMITerGoPZrFCItqSI6rrSio59+vonAMHSPgP3XZAuKymU2mQGlOMjSzdUMQ5+guK4bjuV7qSePY6EwB7yUEYUxojAUoZlHLYhLakYuqZV8DzDhir2J4yrABc0JlDbavaHC912gj9ta5h2s868KNO3q9Auj5-gAYWkYRwWA1pHp4r6vgAMlAAAlOl8EwAAeKVhRPDA83HVVQjecNJCkNhxmSHoOakLmzR5wFeQYCIADcFHQFS0HgRydb1uJgICxI13gc3SjC0AAAMAG0CyLPlXWFABdD3KdlpZqng5B4pSDWwoESBYCYQLzN9z3Mvh0VxQ9k9lSFBy+QkqThGJLrkGVOJ0thp7M9y0BH2YPhy0tDGOBK4M8brH5guLmqsMwPHe0AQ-3QB+Ck2QiNwvE1CJWa+OpoDYA2xckRApZlmU5fQBWy-QZX30-Kv57YQCF4NmNBMhSniXBpQguqTAwuYTTnICu1VNTqRUcnPl9wLEVOsDJXBgQxIrJiaKA7eVoibDmVB5ZeEsliJDYNARAgZMA52OugGSG5dxSCKEsHAplxArk6gKSAbMEjEm0hSAA0pXbMuZfY9AAKwvAHj2Yeo9x6T2nnQK6SkG53XBo9DOHh4DQF0C9N6otxar2lgHDezFYChH2D0E2yASJ9GopUf6twpHA0WLsbsQlnL-BvokbBj9n4qU0ipbSH8v49w4L-BQ-9KH0WAaAq04DqiQJXDAuBsjsDaWQaguIGDziTGwY0UoeDOQEKISQ28ChyGUMQaAWh9FhGQyFDKMREj0AwwYW4yqJZkCOXHNjZaQA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6JgQwFtmBzWsgHsmAN0QtKw9GADa4bDkTRog6JAA08hcCjJESMvlXTIXRFwBGSgDKUA1og1QAZqq4d80mZDLd9AYQ49SABdAF9NBQgsKJw9AyM1DEhiFVJoIg0tKMgiUlQMbwtBQSQOJlDI2Jc3Dy9IAAUg33hA4JCq2JToRGdKAA96ylR1XQALQQQAEydIMaDZ3wrRyCnKGZWAd0p4eFDsiDDskPkjsKA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA)

<details>
<summary>Expand Code</summary>

### Code

```ts
// Playground for #6120
// https://github.com/typescript-eslint/typescript-eslint/issues/6120

//## FAIL
//✅ => Errored as expected

// ✅ Defined here
const myMethod = (props: { MyBoolean: boolean; isMyBoolean: boolean; }) => {
  console.log(props);
};

type MyType = {
  // ✅ Defined here
  MyBoolean: boolean;
  isMyBoolean: boolean;
};

interface SomeObjectType {
  // ✅ Defined here
  something_here: string;
}

const myOwnObjectNotTyped = {
  // ✅ Defined here wihtout types
  MyBoolean: true,
  isMyBoolean: false
};

const someObj = {
  // ✅ Defined here wihtout types
  bad_key: 'hello',
};

//## SHOULD PASS
// 💥 => Should not be Errored

// 💥 Using a defined (external) type, outside programmer's control.
const myOwnObject: MyType = {
  MyBoolean: true,
  isMyBoolean: false
};

// 💥 Using a defined (external) type, outside programmer's control.
const anotherObject: SomeObjectType = {
  // This one should not error, since this format is _required_ for the object
  // to match the type, and so the programmer here has no choice but to use the
  // bad syntax.
  something_here: 'no error!',
};

// 💥 Using a defined (external) type, outside caller's control.
myMethod({ MyBoolean: false, isMyBoolean: true });

// 💥 Using a defined (external) type, outside caller's control.
function doSomething(obj: SomeObjectType) { console.log(obj); }
doSomething({ something_here: 'just works!' });

// 💥 Using a defined (external) type, outside programmer's control.
type CombinedType = SomeObjectType & Record<string, unknown>;
const complex: CombinedType = {
  // Even though CombinedType has a generic `[key: string]` typing, it also
  // explicitly has the key `something_here`, which should be allowed.
  something_here: 'still works',
}

//## To be defined

// 🟡 TO DEFINE
type MapType = Record<string, unknown>;
const someMap: MapType = {
  // The type does not dictate that this key must be snake_case.
  // But without knowing how the Record is parsed, we cannot determine if forcing camelCase is OK.
  bad_key: 5
}

// 🟡 TO DEFINE
function doSomethingAgain(obj: Record<string, any>): void { console.log(obj); }
// The type does not dictate that this key must be snake_case.
// But without knowing how the Record is parsed, we cannot determine if forcing camelCase is OK.
doSomethingAgain({ bad_casing: 'oh no!' });
```

</details>

<details>
<summary>Expand Config</summary>

### Eslint config

```json
{
  "rules": {
    "@typescript-eslint/naming-convention": [
      "error",
      { "selector": "memberLike", "format": ["camelCase"]},
      {
        "selector": "property",
        "types": ["boolean"],
        "format": ["PascalCase"],
        "prefix": ["is", "should", "has", "can", "did", "will"]
      }
    ]
  }
}
```

### TypeScript config

```json
{
  "compilerOptions": {
    "strictNullChecks": true
  }
}
```

| package | version |
| -- | -- |
| `@typescript-eslint/eslint-plugin` | `5.59.11` |
| `@typescript-eslint/parser` | `5.59.11` |
| `TypeScript` | `5.0.4` |
| `ESLint` | `8.39.0` |
| `node` | `web` |
</details>

**Comment by bradzacher** [comment]
> Yet, if the callee typed an argument as a Record<string,any>, my understanding is that the callee intended the keys to be anything; otherwise it should have been typed with an object literal.

If there are no types for the keys because they are typed via a record then the rule should default to using whatever the config is for the entire codebase. Eg for most codebases it would be camelCase.

This is the same as if you had an untyped object literal like `const x = {...}`

> otherwise it should have been typed with an object literal.

This isn't a correct line of thought - it's very common to have cases like config keys that don't have any pre-defined shape, but you want a standard naming convention for them.

ESLint rules are a funny case. You mention "they should be kebab-case" but that's just a general agreed upon convention in the ecosystem - there's no constraint that forces them to be kebab-case!

**Comment by sharky98** [comment]
> > Yet, if the callee typed an argument as a Record<string,any>, my understanding is that the callee intended the keys to be anything; otherwise it should have been typed with an object literal.
> 
> If there are no types for the keys because they are typed via a record then the rule should default to using whatever the config is for the entire codebase. Eg for most codebases it would be camelCase.
> 
> This is the same as if you had an untyped object literal like `const x = {...}` - there's no type specified but you still want it to follow the codebase standard.

I agree with this: `rule should default to using whatever the config is for the entire codebase`. The thing is, as the caller of an external function, I have no control on what rule the function codebase is using. Even if `camelCase` is the default for most codebases, I think it is a false assumption for a rule to make that *all* codebase would be `camelCase` and force that into an object that me, the programmer, have no control over; even if they are `Record<>`.
 
> > otherwise it should have been typed with an object literal.
> 
> This isn't a correct line of thought - it's very common to have cases like config keys that don't have any pre-defined shape, but you want a standard naming convention for them.

I agree here also. This is like I said in, perhaps less clear way just below: `For sure, there would be rules in places on how the called would parse the keys, noted in the documentation somewhere.` Obviously config keys are the typical example for this (like the ESLint rules names 😉).

When I said this part about being typed with an object literal, I was referring to types that could have a lot of items, but are fixed and defined (basically just the programmer using a lazy `Record<>` instead of spelling it out.

However, the point I was trying to make here is again, me, as the programmer using an external module, I have absolutely no control on what standard naming convention the type I am using have in its codebase. I could be forced to use `kebab-case`, `snake_case` or the like. As such, since the `ObjectExpression` is typed, no matter what the typing says, `naming-convention` should not flag my properties in error.

> ESLint rules are a funny case. You mention "they should be kebab-case" but that's just a general agreed upon convention in the ecosystem - there's no constraint that forces them to be kebab-case!

I agree, there is no constraint that forces the **creator** of a ESLint rules to be `kebab-case`. Yet, me, as the programmer, I am forced to use that `kebab-case` convention. As such, if I would have a `naming-convention` forcing `camelCase` and my ESLint config was done in TypeScript, given the previous discussion regarding the `Record<>`, the rule would trigger an error!

**Comment by bradzacher** [comment]
Everything you're describing really makes it sound like you just don't want any enforcement of property naming conventions in your codebase because either:
- a property is strictly typed already and that name is enforced by the library's convention or by `naming-convention` on the type, or
- a property has no type at all (eg record) and thus in your opinion has no convention to specify.

So really in your codebase you don't want this feature - you just want to set `camelCase` for type properties and disable validation of object properties altogether.

**Comment by sharky98** [comment]
Maybe I am expressing myself wrongly; English isn't my first language... And I have known issues to express clearly my though in writing (or verbally for that matter!).

I would like to have an enforcement of a naming convention on the properties on any object that is generated and typed (or not) as part of my own codebase.

However, I do not want it to be applied when my own codebase is not responsible of selecting the naming convention for that particular object.

I fail to see how any given codebase would want to have linting errors on naming convention outside of its own control; which is the alternative (well more like status quo) I see proposed here.

Maybe disabling validation of object properties altogether is the easy solution. I am just wondering if there is situation where the naming convention should have applied but was not due to disabling the validation.

----
Bit of brainstorming here for myself regarding that solution (disable `objectLiteralProperty ` altogether), while making sure all case would get a naming-convention nonetheless.

* Initialization of a Variable Declaration.
   * Internal variable with strictly defined type/interface.
      * ✅ Naming Convention would be dealt by the `typeProperty` selector.
   * Internal variable with generic defined type/interface (record or `[key: string]`).
      * ❌ Naming Convention cannot be applied to the `typeProperty` selector.
   * External strictly defined type.
      * ✅ OK to disable, we are forced to use external convention.
   * External generic defined type (record or `[key: string]`).
      * ✅ OK to disable, we are forced to use external convention.
* Argument of a Call Expression.
   * Internal function with argument that are strictly defined with type/interface.
      * ✅ Naming Convention would be dealt by the `parameterProperty` selector.
   * Internal function with argument containing generic defined type/interface (record or `[key: string]`).
      * ❌ Naming Convention cannot be applied to the `parameterProperty` selector.
   * External strictly defined type.
      * ✅ OK to disable, we are forced to use external convention.
   * External generic defined type (record or `[key: string]`).
      * ✅ OK to disable, we are forced to use external convention.

We are focusing on `objectLiteralProperty` selector, but I wonder if other item could have the same issue (`classProperty`, `enumMember`, `parameterProperty`?). Need to check if calling a method or a function externally defined triggers the naming-convention rule. Quick check seems to point that no it does not.

**Comment by bradzacher** [comment]
No, I fully understood what you're saying. The problem is that you're making a delineation between "internal" and "external" types - but in reality they are the same thing. It doesn't matter where a record type is defined - it's just a record type that specifies keys with no structure.

For example you could envision that an "internal" type is specified as a record that is intended to be passed to an "external" location that accepts a record. In that case your delineation between the two falls over because you would want that object to be typed based on that "external" expectation, even though its an "internal" type.

The distinction you've declared is arbitrary based on how a human might draw lines and categorise things - which isn't something that can be encoded.

Refardless though - if there is some usecase that an API specifies that it accepts a record and it specifically really cares about the format of the keys (I haven't seen it, personally) then there's a super-duper rare edge-case that disable comments can be used.
Otherwise keys can and should match the codebase style if they are untyped.

There's a lot of other cases you haven't mentioned - for example passing an object toa location expecting  `any`/`unknown`/`object`/`{}`.

The solution to this issue can be boils down to the following: "if a property is assigned to a typed location where that propertu name exists on the type, then the property is ignored, otherwise it uses the codebase's casing".

> but I wonder if other item could have the same issue (classProperty, enumMember, parameterProperty?)

Enum members - nope as there's no way to define a type interface for an enum.

Class properties - yes, in a way, as there may be an interface the class needs to adhere to.
However as mentioned - you're getting into SUPER weird and probably non-existent usecases there. Theoretically yes; realistically no.

**Comment by sharky98** [comment]
> if there is some usecase that an API specifies that it accepts a record and it specifically really cares about the format of the keys (I haven't seen it, personally) then there's a super-duper rare edge-case

> you're getting into SUPER weird and probably non-existent usecases there.

You're describing my life right there. My brain is wired in such a way that I ONLY think of those cases first. 😂 Thanks for putting me back to my place!

Anyhow, ignoring the `Record<>` thing, all other uses cases (I think) I have could fall into the *disable `objectLiteralProperty`*.

**Comment by gauntface** [comment]
I could have sworn a long time ago that `objectLiteralProperty` would be linted with the caveat that strings would not need to be formatted. This worked great for a number of cases and might satisfy this issue.

```
const example = {
   // CamelCase is the default
   thisIsOK: true,

   // Other styles will fail
   snake_case_is_not_ok: false,
   PascalCaseIsNotOK: false,

   // But anything in a string was ok
   'This is OK': true,
   'this_is_ok_too': true,
   'ThisIsAlsoOK': true,
};
```

Without this escape hatch, it's more tempting to disable the rule for `objectLiteralProperties` than try replace the default config with something to account for this.

**Comment by bradzacher** [comment]
> I could have sworn a long time ago that objectLiteralProperty would be linted with the caveat that strings would not need to be formatted.

A long time ago it did - then an overwhelming number of people asked for this to change because people would accidentally (or purposely) quote props which would allow them to use invalid names for the  codebase's conventions. So the rule thus was changed to match quotes properties too.

The rule was augmented to include the [`requiresQuotes`](https://typescript-eslint.io/rules/naming-convention/#selector-options:~:text=requiresQuotes%20%2D%20matches%20any%20name%20that%20requires%20quotes%20as%20it%20is%20not%20a%20valid%20identifier%20(i.e.%20has%20a%20space%2C%20a%20dash%2C%20etc%20in%20it).) modifier which allows you to match and either enforce or exclude properties which specifically *_require_* quoting (eg the name includes spaces).

In general you should either add exceptions for cases you want to allow, or use disable comments to one-off annotate cases where you are purposely circumventing the rule.

**Comment by JoshuaKGoldberg** [comment]
Per #8792, we're feature freezing this rule. It's reached the point of complexity where we're having a hard time justifying working on it. 

If you'd like to see this option happen, I'd encourage you to add it to a community project such as [eslint.style](https://eslint.style). Cheers! 💙

**Comment by karlhorky** [comment]
I also mentioned this just now over in my question to the ESLint Stylistic maintainers about potentially migrating this rule to ESLint Stylistic:

- https://github.com/typescript-eslint/typescript-eslint/issues/8792#issuecomment-2025344673
